isis/src/system/objs/KernelDb/KernelDb.cpp

```diff
@@ -581,20 +582,34 @@ namespace Isis {
     bool matchTime = !grp.hasKeyword("Time");
     bool matchKeywords = true;
 
+    // If the kernel segment has an instrument specification that doesn't match
+    // the instrument id in the label, don't perform the rest of the checks
+    if (grp.hasKeyword("Instrument")){
+        const PvlGroup &inst = lab.findGroup("Instrument", Pvl::Traverse);
+        if (inst["InstrumentId"][0].toStdString() != grp["Instrument"][0].toStdString()){
+          return false;
+        };
+    }
+
     // First, the time search. Loop through the keywords, if the name isn't
     //  Time then skip it. If it is, then get the start/end times and keep
     //  looking until one is found.
     for (int keyword = 0; keyword < grp.keywords(); keyword++) {
       PvlKeyword key = grp[keyword];
 
       if (key.isNamed("Time")) {
+        double offset = 0;
         // Pull the selections start and end time out
         iTime kernelStart = (QString) key[0];
         iTime kernelEnd   = (QString) key[1];
 
+        if (grp.hasKeyword("TimeOffset")){
+          offset = (double) grp["TimeOffset"];
+        }
+
         // If the kernel times inside of the requested times we
         // set the matchTime to be true.
-        if ((kernelStart <= timeToMatch) && (kernelEnd >= timeToMatch)) {
+        if ((kernelStart - offset <= timeToMatch) && (kernelEnd + offset >= timeToMatch)) {
```

Comment:
I'm not sure if this is the correct logic for the offset.  My understanding is that the offset is something like abs(label_start - camera_start), so we're just expanding the search window by that amount in each direction.

Reply:
This is not completely correct. This does not have my changes in which there is an offset for start time and an offset for end time. "TimeOffset" keyword no longer exists in dbs. It's now StartOffset and EndOffset. There also needs to be checks to see if both exist. It's possible for only StartTime offset to exist or just EndTime, or both.